src/python/grpcio_tests/tests_aio/health_check/health_servicer_test.py

```diff
@@ -0,0 +1,230 @@
+
+    async def setUp(self):
+        self._servicer = health.AsyncHealthServicer()
+        await self._servicer.set('', health_pb2.HealthCheckResponse.SERVING)
+        await self._servicer.set(_SERVING_SERVICE,
+                                 health_pb2.HealthCheckResponse.SERVING)
+        await self._servicer.set(_UNKNOWN_SERVICE,
+                                 health_pb2.HealthCheckResponse.UNKNOWN)
+        await self._servicer.set(_NOT_SERVING_SERVICE,
+                                 health_pb2.HealthCheckResponse.NOT_SERVING)
+        self._server = aio.server()
+        port = self._server.add_insecure_port('[::]:0')
+        health_pb2_grpc.add_HealthServicer_to_server(self._servicer,
+                                                     self._server)
+        await self._server.start()
+
+        self._channel = aio.insecure_channel('localhost:%d' % port)
+        self._stub = health_pb2_grpc.HealthStub(self._channel)
+
+    async def tearDown(self):
+        await self._channel.close()
+        await self._server.stop(None)
+
+    async def test_check_empty_service(self):
+        request = health_pb2.HealthCheckRequest()
+        resp = await self._stub.Check(request)
+        self.assertEqual(health_pb2.HealthCheckResponse.SERVING, resp.status)
+
+    async def test_check_serving_service(self):
+        request = health_pb2.HealthCheckRequest(service=_SERVING_SERVICE)
+        resp = await self._stub.Check(request)
+        self.assertEqual(health_pb2.HealthCheckResponse.SERVING, resp.status)
+
+    async def test_check_unknown_service(self):
+        request = health_pb2.HealthCheckRequest(service=_UNKNOWN_SERVICE)
+        resp = await self._stub.Check(request)
+        self.assertEqual(health_pb2.HealthCheckResponse.UNKNOWN, resp.status)
+
+    async def test_check_not_serving_service(self):
+        request = health_pb2.HealthCheckRequest(service=_NOT_SERVING_SERVICE)
+        resp = await self._stub.Check(request)
+        self.assertEqual(health_pb2.HealthCheckResponse.NOT_SERVING,
+                         resp.status)
+
+    async def test_check_not_found_service(self):
+        request = health_pb2.HealthCheckRequest(service='not-found')
+        with self.assertRaises(aio.AioRpcError) as context:
+            await self._stub.Check(request)
+
+        self.assertEqual(grpc.StatusCode.NOT_FOUND, context.exception.code())
+
+    async def test_health_service_name(self):
+        self.assertEqual(health.SERVICE_NAME, 'grpc.health.v1.Health')
+
+    async def test_watch_empty_service(self):
+        request = health_pb2.HealthCheckRequest(service='')
+
+        call = self._stub.Watch(request)
+        queue = asyncio.Queue()
+        task = self.loop.create_task(_pipe_to_queue(call, queue))
+
+        self.assertEqual(health_pb2.HealthCheckResponse.SERVING,
+                         (await queue.get()).status)
+
+        call.cancel()
+        await task
+        self.assertTrue(queue.empty())
```

Comment:
If the points is that once the request is finished no more messages should be seen, would it be better to create a test for this and not having to repeat in all of the tests the `self.assertTrue(queue.empty())` ?

Reply:
This test is guarding that the `Watch` method can properly fetch status of the overall health status. If we switch to `Check`, then it provides no value.

The `self.assertTrue(queue.empty())` is essential in each scenarios that it ensures no operation to the servicer will cause additional notification.

Reply:
yes is about the last one

> The self.assertTrue(queue.empty()) is essential in each scenarios that it ensures no operation to the servicer will cause additional notification.

Could we cover this test case in a unique test and not having to repeat the same assert in all of the other tests?

Reply:
Adding one test case that runs all seven scenarios might be too complicated? I felt like keep them separate improves readability, since each one is nice and short.